### DISCUSSION TITLE
Error: refusing to allow a GitHub App to create or update workflow

### DISCUSSION THREAD

**Discussion by lktslionel** [community/community#35410] [Actions]
Hello,
I have a weird behaviour on GitHub workflow. I have a workflow that:

1.  checkout the code at a specific tag 
2. Make some modifications
3. Create a branch locally 
4. Tag the branch
5. Push all changes with tag to GitHub 

And i got the following error: 

```
To https://github.com/ORG/REPO
 ! [remote rejected] release/v3.7.0 -> release/v3.7.0 (refusing to allow a GitHub App to create or update workflow `.github/workflows/deliver.yml` without `workflows` permission)
error: failed to push some refs to 'https://github.com/ORG/REPO'
```

I don't understand why the workflow think I'm (self) updating the current workflow? 
It worked fine before but now it start failing on all my workflows.

Below is the workflow

```yaml

# deliver.yml
name: Delivery Workflow
on:
  workflow_dispatch:
    inputs:
      component-version:
        description: 'Component version you want to deliver. Eg: X.Y.z-rc.N'
        required: true
        type: string

jobs:
  deliver:
    name: Deliver
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: 'v${{ env.COMPONENT_VERSION }}'

      - name: Config Git and NPM
        run: |
          git config user.name github-actions
          git config user.email github-actions@github.com
          npm config set user github-actions
          npm config set email github-actions@github.com

      - name: Bump version
        run: |
          npm version v2.0.0 # v2.0.0 is an example

      - name: Commit changes
        run: |
          git push origin v2.0.0 # v2.0.0 is an example

```

> I don't use a PAT and I want to not have to.

Any idea?
Thanks

**Comment by kammce** [comment]
I'm having this issue as well

**Comment by pie972** [comment]
You need to grant the workflows permission to your GitHub App. 
1. Go to the settings page of your GitHub App.
2. Under "Permissions & events", click on "Configure permissions".
3. In the "Repository permissions" section, check the "Read & write" permission for the "Workflows" option.
4. Click on "Save changes".

**Comment by tempestrock** [comment]
For those who come across this issue here:
I had the same problem. The usual answer I read everywhere was to add `token: ${{ secrets.GITHUB_TOKEN }}` in the `with` section of the checkout action. However that didn't work for me.
The solution for me was the following: My workflow got triggered by a `repository_dispatch` event that came from a different repository. And in that case the `GITHUB_TOKEN` in the `with` section of the checkout action was not enough. I had to create a personal access token (PAT), give it the permission `workflow`, and put that into the `with` section.

**Reply by lucasresck** [reply]
It may not be obvious that, after creating a personal access token (PAT) at [the settings](https://github.com/settings/tokens?type=beta) and giving it `workflow` permissions, it is necessary to copy this key and add it as a secret at the repository's secrets (Repo > Settings > Secrets and variables > Actions). Only after that it is possible to use this new repository's secret, which is a PAT, within the `with` section. For instance, if your secret is named `WORKFLOW_TOKEN`, one may use it with
```
with:
    token: ${{ secrets.WORKFLOW_TOKEN }}
```

**Reply by liquidhiter** [reply]
What do you mean by `copy this key`? I have followed your suggestions to create a PAT with workflow permissions. However, I didn't figure out how to add the PAT as a secret in the repo settings. I have searched on Google and read the GitHub documentation, but still didn't find anything. Can you please shed some light on this?

**Reply by bewuethr** [reply]
Secrets are under Setttings > Secrets and Variables > Actions, URL `https://github.com/OWNER/REPO/settings/secrets/actions`

**Reply by liquidhiter** [reply]
Thanks! I think I didn't clearly describe the issue I am dealing with now. According to `lucasresck`'s suggestion, I know I can add the PAT as a secret `at the repository's secrets (Repo > Settings > Secrets and variables > Actions).`

But I don't know what needs to be added in the `Repository secrets`:
- I think I need to add a new repository secret by clicking `New repository secret`
- I know the name of this new secret is `WORKFLOW_TOKEN` (it is the name of the new PAT with workflow permission). But I am asked to also enter the value, which value should I use? I carefully checked the newly created PAT, but couldn't find any value.

- actions secrets option
![image](https://github.com/community/community/assets/35713945/091b177b-66d2-4811-839e-9c16d86bd328)

- new PAT 
![image](https://github.com/community/community/assets/35713945/9ffcaade-fee4-4452-8e1a-a4e98375d7f3)

Am I doing something wrong?

**Reply by lucasresck** [reply]
Hi, @liquidhiter!

When you create a PAT in GitHub settings, it allows you to copy the token:

![image](https://github.com/community/community/assets/41991486/9a15f5fe-2642-4f21-9053-4ba625e8a434)

This information will not reappear after leaving the page. This is the information you will insert under "Secret" in your GitHub repo settings.

If you didn't copy or save the token, you will have to either regenerate the token or create another token. In your last screenshot, just press "Regenerate token" and copy this value. Then insert this into your repo secrets.

For clarification, I don't think PAT and secret need to have the same name `WORKFLOW_TOKEN`. They both having the same name is just for organization. What needs to have the same name are the repo secret and the token in the `.yml` of your action. The "link" between your secret and the PAT is the token itself.

Basically, when you do all this stuff, your action will ask for a `secrets.WORKFLOW_TOKEN`, which is saved in your repo secrets under the name `WORKFLOW`, and the token associated to this secret can be used for authentication because it is a personal access token (PAT) with workflow permissions (the PAT's name doesn't matter).

There are some months I don't work with these matters. Sorry for any inaccuracy. Let me know!

**Comment by ghost8658private** [comment]
Hey @lktslionel have you found a solution to this?? I've been stuck here for a week now, I tried different things and none seems to work

**Reply by lktslionel** [reply]
No. The PAT v1 works but not the PAT v2.
So weird

**Comment by maboloshi** [comment]
_**Sorry for the misinterpretation, but the following is only for permissions that use tokens generated automatically by GitHub Action.**_

In the workflow file, add:
```yaml
# Set permissions on GITHUB_TOKEN to allow updates to GitHub Actions workflows
permissions:
  actions: write
```

In fact, you can also modify the following default parameters：
```yaml
permissions:
  actions: read
  checks: read
  contents: read
  deployments: read
  discussions: read
  issues: read
  metadata: read
  pages: read
  packages: read
  pull-requests: read
  repository-projects: read
  security-events: read
  statuses: read
```
see：
- https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token
- https://docs.github.com/zh/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories

**Reply by maboloshi** [reply]
The installation access token permissions for GitHub App, in addition to referencing [here](https://github.com/orgs/community/discussions/35410#discussioncomment-5427173), you'll need to go to the Apps page in the repository settings to confirm authorization updates.

**Reply by Zero-Fanker** [reply]
This actually works for me, no idea why so many down votes.
But referring to Github App setting is a bit confusing, since it controls only the 'default permission' now, which means it will still be overriden during the workflow run with proper permission declaration as you demonstrated above.

**Comment by bewuethr** [comment]
This is pretty high up when searching for the error message, and some of the answers are outdated. As of today (November 2023), I think this is correct:

- `GITHUB_TOKEN` can't be given permissions to modify workflow files. `actions: write` is not sufficient. To modify a workflow file from an actions workflow, you must use a personal access token.
- Fine-grained PATs now work with GraphQL, so they can be used. To make changes to a workflow file and commit it, these repo scopes are required: `contents:write`, `workflows:write`. `metadata:read` is set automatically.
- To authenticate the `git` command line client, use the `token` option on `checkout`

Something like this:

```yaml
jobs:
  changewf:
    name: Change a workflow file
    runs-on: ubuntu-22.04
    permissions:
      contents: write
      pull-requests: write

    steps:
      - name: Check out repository
        uses: actions/checkout@v4.1.1
        with:
          # Fine-grained PAT with contents:write and workflows:write
          # scopes
          token: ${{ secrets.WORKFLOW_TOKEN }}

      - name: Make change to workflow file
          GITHUB_TOKEN: ${{ github.token }}
        run: |
          # Create new branch
          git switch -c "feature"

          # Edit workflow file
          echo "# I am a change!" >> .github/workflows/workflow.yml

          git config --global user.name "github-actions"
          git config --global user.email \
              "41898282+github-actions[bot]@users.noreply.github.com"

          git add .github/workflows/workflow.yml
          git commit --message "Update workflow"

          # This authenticates with WORKFLOW_TOKEN, because it was
          # used with the checkout action
          git push --set-upstream origin "feature"
          
          # This authenticates with GITHUB_TOKEN, using the scopes
          # set in jobs.changewf.permissions
          gh pr create --title "Update workflow" --body ''
```

**Reply by iamutkarshtiwari** [reply]
@bewuethr Thank you so much for elaborating on the solution. You are a savior! 🙏

**Reply by hareacc** [reply]
@bewuethr 

I did same like as per your sample workflow but getting error _remote: Write access to repository not granted._ during checkout step

When I create Fine grade PAT, it does not allow me to select my organization's repository so I selected all but after creating token, it show "_This token has access to all repositories owned by you._",  here it say owned by you but workflow I tried is in my organization repo, does that create issue?

![image](https://github.com/community/community/assets/115476942/64ec8a2e-1bc8-4ca7-bc28-0232569e17e7)

**Reply by bewuethr** [reply]
Right above that, it should show you the organization, like this:

![image](https://github.com/community/community/assets/8521043/8041abb9-536f-4f45-9a64-2057cdc48ebb)

Not sure what might prevent that from happening, maybe SSO settings or similar?

**Reply by hareacc** [reply]
There are no SSO. Also after Access on, it showing me my GH Id.

Thanks for your quick response. @bewuethr

**Reply by devnoname120** [reply]
@hareacc Create a classic token instead of a fine-grained token.

**Comment by dewodt** [comment]
For me, it turns out that my CI (in my case prettier) was trying to modify other `.github/workflows` files, so if that the case for you, try ignoring `.github` folder in `.prettierignore`.

**Reply by rdg-develoci** [reply]
Thanks man!! That was my issue as well.

**Reply by teloryfrozy** [reply]
Anyone having issues and not wanting to create a PAT ☝️
You can either put: ".github/*" in a .prettierignore file

If using npm as CI tool to lint your code:
Simply set npm format command to: "prettier --write . --config .prettierrc.json '!**/*.{yaml,yml}'" in package.json (adapt config path)

**Comment by kevanantha** [comment]
If you are using Github CLI, you need to re-auth with the [correct scope](https://cli.github.com/manual/gh_auth_login) -> `gh auth login -s workflow` and you are good to go.

Your token scopes should be like this `'gist', 'read:org', 'repo', 'workflow'` when you run command `gh auth status`

**Reply by monzim** [reply]
worked for me thanks

**Reply by hrideshmg** [reply]
Worked here as well, pretty weird why this is coming up now. I have modified workflows in the past without this specific scope.

**Comment by louwers** [comment]
I am not even modifying a workflow file, yet I still get this.

**Reply by marco-merola-ef** [reply]
This is happening for me as well. Did you understand why?

It works if I do:
 `git branch release/v1.0`
 `git checkout release/v1.0`
 `git push origin release/v1.0`

It throws the error if I do:
 `git branch release/v1.0 a40f08f31c783b96b5678a802901cba961366f28`
 `git checkout release/v1.0`
 `git push origin release/v1.0`

I don't understand why adding the commit SHA makes github thinking that I am modifying the workflow?!

**Reply by louwers** [reply]
Yeah this error seems to be dependent on the branch name...

I also included a git SHA in the branch name!!

**Reply by jduan-highnote** [reply]
Same. In my case, I'm updating a tag instead. I simply do this:
```
            git tag dev-deploy origin/main
            git push --tags --force
```

I want to update the `dev-deploy` tag to point to the latest origin/main. That somehow makes github actions think I'm updating some workflow file. What's even weirder is that every time it fails, it shows some "random" workflow file.

**Reply by carlcsaposs-canonical** [reply]
I believe this might be the root cause: https://github.com/orgs/community/discussions/151442

**Reply by LianwMS** [reply]
@jduan-highnote , do you have any solution to fix? I meet the same error when I try to update the tag

**Reply by marco-merola-ef** [reply]
Try doing everything in the push like this:
`git push origin <commit-sha>:refs/tags/<tag-name>`
Example:
`git push origin a1b2c3d4:refs/tags/dev-deploy`

**Reply by carlcsaposs-canonical** [reply]
@marco-merola-ef I tried replacing `git push origin <tag-name>` with `git push origin <commit-sha>:refs/tags/<tag-name>` in the steps to reproduce for https://github.com/orgs/community/discussions/151442 and got the same error: https://github.com/carlcsaposs-canonical/test-fix1-bug-report-github-actions-tags/actions/runs/16264203666/job/45916110377#step:3:8

**Reply by marco-merola-ef** [reply]
Please remove the first line. If your intent is to create and push a tag at a specific commit sha, there is no need to create it locally.
And I think that's reason why github is complaining.
`git tag candidate HEAD^ --force`

**Reply by carlcsaposs-canonical** [reply]
Just tried it; looks like it's failing with the same error: https://github.com/carlcsaposs-canonical/test-fix1-bug-report-github-actions-tags/actions/runs/16320254822/job/46095907720

**Reply by marco-merola-ef** [reply]
I guess it's because the commit `635ff6fbfe1cea957f9f162aefbe4ad6e209ebd3` contains changes to .github/workflows/*.yaml
Try tagging a commit that doesn't contain any changes to that folder. GitHub enforces workflow file update permissions even for tag pushes if the commit contains workflow changes.
The alternative otherwise is to add the `workflows` permission as requested in the error.

**Reply by carlcsaposs-canonical** [reply]
Here's a commit that doesn't change workflows that fails with the same issue: https://github.com/carlcsaposs-canonical/test-fix1-bug-report-github-actions-tags/actions/runs/16367121336/job/46246796192

Our understanding of the exact root cause (which we believe is an issue that should be fixed) is here: https://github.com/orgs/community/discussions/151442

**Comment by JeanBarriere** [comment]
I had the same issue and I really do not want to create a PAT for this since I know the default GITHUB_TOKEN can create PRs (so obviously it should be able to create branches, right?!).
I found a solution and it's actually a very smart behaviour from GitHub: the error happens because the branch contains changes in 
`.github/workflows` compared to the default branch (e.g. `main`).

To solve this, you can checkout the default branches files into your new branch, commit those changes and 🎉 , you're synced and allowed to push!

It looks like this:
```sh
git checkout $COMMIT_SHA
git switch --create $BRANCH_NAME
git checkout origin/main -- ${ROOT_DIR}/.github/workflows
git add ${ROOT_DIR}/.github/workflows
git commit -m "chore: checkout .github/workflows files from main branch"
git push origin $BRANCH_NAME
```

Hope this will help!

**Reply by JJPowell** [reply]
This was the solution that worked for me without having to create a PAT.

**Reply by boholder** [reply]
The commands are pasting workflow content on origin/`main` branch to origin/`$BRANCH_NAME`. Not sure if it works for "merging PR branches into `main`" situation.

**Reply by ryanpeach** [reply]
Yeah but I hit this error because I DO want my github actions event to edit the workflows file.

**Reply by boolYikes** [reply]
Thanks I really did not want to involve PAT 👍

**Reply by ArchieMeng** [reply]
> Yeah but I hit this error because I DO want my github actions event to edit the workflows file.

Exactly. If someone just wanted to manually update the branch, then this question should never be asked.

**Comment by paololazzari** [comment]
I am working with a github app and I just ran into this issue as well (the commit itself did not modify any workflow file).
This was my code:

```
- name: Generate a token
  id: generate-token
  uses: actions/create-github-app-token@v1
  with:
    app-id: ${{ secrets.APP_ID }}
    private-key: ${{ secrets.PRIVATE_KEY }}

- name: 'Checkout'
  uses: actions/checkout@v3
  with:
    token: ${{ steps.generate-token.outputs.token }}

- name: 'foo'
  run: |-
    ... # make changes

- name: 'Open pull request'
  env:
    GH_TOKEN: ${{ steps.generate-token.outputs.token }}
  run: |-
    # Setup git email and username
    git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
    git config user.name "github-actions[bot]"
    git config --global "url.https://${GH_TOKEN}@github.com/.insteadOf" https://github.com/

    git add ...
    git commit -m ...
    git push origin ...

    gh pr create ...
```

The issue turned out to be due to the following line of code:

```
git config --global "url.https://${GH_TOKEN}@github.com/.insteadOf" https://github.com/
```

by removing it everything worked fine.

**Reply by bewuethr** [reply]
Yeah, checkout persists the auth token in the local Git config. It took me a long time to figure that out when I needed a Git CLI client with different permissions, but all it takes is setting a different `token` on `checkout`.

**Comment by mietzen** [comment]
This worked for me:

- Create a Github app with permissions for:
  - workflow: read and write 
  - content: read and write 
- Install it into your account and give it permission to your repo
- Add private key and app id as secrets

For more info read docs of: https://github.com/actions/create-github-app-token

I wanted to sync a fork including tags:

```yaml
name: Sync Fork

on:
  schedule:
    - cron: '0 2 * * *'
  workflow_dispatch: # on button click
  
jobs:
  sync:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/create-github-app-token@v1
        id: generate-token
        with:
          app-id: ${{ secrets.APP_ID }}
          private-key: ${{ secrets.APP_PRIVATE_KEY }}
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0
          fetch-tags: true
          token: ${{ steps.generate-token.outputs.token }}
      - name: Merge upstream
        run: |          
          git remote add upstream https://github.com/${FORK-PARENT}
          git fetch upstream
          git fetch --tags upstream
          
          git merge --no-edit upstream/master
          git push
          git push --tags
```

**Reply by Yevelkin** [reply]
This worked for me with changing workflows files via Github app!
Thank you :)

**Comment by GoncaloPT** [comment]
For me, it was this dreaded permission update of the app.
In the organization page, where you manage the Apps and can set their permissions, the following message is showed near the save:
<img width="864" alt="image" src="https://github.com/user-attachments/assets/f53088fe-5c2d-4eba-a259-22c68877ef84" />

This means that you must go, repo by repo, and accept the permission change.

**Reply by devnoname120** [reply]
You can just create a classic Personal Access Token and use that instead of the default `GITHUB_TOKEN`

**Reply by Hansemeister** [reply]
This was the fix for me. I had to contact our organization owner and have them approve the change to add the 'workflows' permission to the GitHub App.
If that red note is stil there on a browser refresh, means that it still hasn't been approved.

**Comment by innovara** [comment]
Just to add to the number of solutions here, this is what I recently did for an action that would rebase a fork from upstream adding a workflow in the process.

- Created a fine-grained Personal access token with read access to metadata and read-write access to code and workflows for that repository only. Nothing else.
- Added an Action repository secret named `WORKFLOW_TOKEN` with the fine-grained PAT from the previous step.
- Added the following action to my repo:

```name: Daily Upstream Rebase
on:
  schedule:
  - cron: "0 0 * * *"
  workflow_dispatch:
    inputs:
      upstream:
        description: <user>/<repo> or the full HTTP URL
        required: false
      branch:
        description: The upstream branch that is rebased on
        required: false
        default:  master
      depth:
        description: Greater than the number of commits the upstream made in a period
        required: false
        default:  100
      push:
        description: Do the force push in this action
        required: false
        default:  true

jobs:
  sync:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@master
      with:
        fetch-depth: 10  # greater than the number of commits you made
        persist-credentials: false
    - shell: bash
      env:
        WORKFLOW_TOKEN: ${{ secrets.WORKFLOW_TOKEN }}
      run: |
        set -ex;
        UPSTREAM=${{ inputs.upstream }};
        if [ -z $UPSTREAM ]; then
            echo $WORKFLOW_TOKEN | gh auth login --with-token;
            UPSTREAM=$(gh api repos/:owner/:repo --jq .parent.full_name);
            if [ -z $UPSTREAM ]; then echo "Can't find upstream" >&2 && exit 1; fi;
        fi;
        if [ ! $(echo $UPSTREAM | egrep '^(http|git@)') ]; then
            UPSTREAM=https://github.com/$UPSTREAM.git
        fi;
        if [ ${{ inputs.depth || 100 }} -ne 0 ]; then
            DEPTH=--depth=${{ inputs.depth || 100 }}
        fi;
        git remote add upstream $UPSTREAM;
        git fetch upstream ${{ inputs.branch || 'master' }} $DEPTH;
        git config --local user.email "41898282+github-actions[bot]@users.noreply.github.com";
        git config --local user.name  "GitHub Actions";
        git rebase upstream/${{ inputs.branch || 'master' }};
        if [ "${{ inputs.push || true }}" = "true" -a "$(git status | grep diverged)" ]; then
            ORIGIN_URL=$(git remote get-url origin);
            git push "${ORIGIN_URL/https:\/\//https:\/\/$WORKFLOW_TOKEN@}" $(git branch --show-current) --force;
        fi;
```

One thing that you shouldn't overlook is that you need to add `persist-credentials: false` to `actions/checkout@master` because `token` defaults to `${{ github.token }}` on that action, and if stored it will be subsequently used for `git push`, leading to the error regarding insufficient permissions to create or update workflow.

**Comment by kurdin** [comment]
I had the same problem using organization's repo and secrets.GITHUB_TOKEN, the fix was simply to add  `actions: write` to permissions section of the workflow file.

permissions:
  contents: write
  **actions: write**
  pull-requests: write
  packages: write

**Comment by ArchieMeng** [comment]
Hi everyone,

Just wanted to confirm that the solution proposed by @mietzen in this comment: https://github.com/orgs/community/discussions/35410#discussioncomment-10701448 using `actions/create-github-app-token` works perfectly for me to get the necessary permissions within a workflow triggered by a GitHub App's `repository_dispatch` event.

I've found a few additional points that might be helpful for others, based on my experience:

*   **Other Common Solutions Were Ineffective:** For my specific scenario, I tried various other approaches to grant elevated permissions, including adding `actions: write` permissions directly to the workflow YAML and granting extensive `actions` or `workflow` permissions to a Personal Access Token (PAT). None of these methods worked for me, which is why the GitHub App approach became essential for achieving the required access.

*   **No Additional Workflow-Level Permissions Required:** You generally don't need to grant additional `permissions` at the workflow YAML level (e.g., `permissions: contents: write`) *as long as* the GitHub App itself has been granted the required `contents` and `workflows` write permissions (or whatever specific permissions your workflow needs) during its installation. The token generated by `actions/create-github-app-token` will inherit the App's permissions.

*   **App Installation is Crucial:** Ensure that your GitHub App is correctly installed on the repository where the workflow is running. This is a fundamental requirement for `actions/create-github-app-token` to function and generate a valid token.

*   **Checkout with the Generated Token:** When checking out the repository, you *must* use the token generated by `actions/create-github-app-token` for any operations requiring write access. For example:

    ```yaml
    - name: Generate GitHub App Token
      id: generate_token
      uses: actions/create-github-app-token@v2
      with:
        app-id: ${{ secrets.APP_ID }} # Or your app ID
        private-key: ${{ secrets.PRIVATE_KEY }} # Or your private key

    - name: Checkout repository
      uses: actions/checkout@v4
      with:
        token: ${{ steps.generate_token.outputs.token }}
    ```
    (While `persist-credentials: false` is an option for `actions/checkout`, using the generated token directly as shown above is the most straightforward way to ensure your subsequent `git` commands have the correct authentication.)

Hope this helps others facing similar permission challenges!

**Comment by jefferywright359-eng** [comment]
This error typically appears if your push includes changes that GitHub interprets as modifying a workflow file, and the app/token no longer has workflows scope; it’s similar to [luxury pet transportation](https://url-shortener.me/79DP) where the process seems unchanged, but stricter checks suddenly apply without obvious changes on your end.
